### ISSUE TITLE
emb v 2

### ISSUE THREAD

**Issue by SkBlaz** [SkBlaz/py3plex#1272]
Refactor and polish the embedding subsystem so embeddings behave as a native first-class primitive across the entire py3plex API surface: DSL (Q), graph_ops, pipelines, and data export layers.

Goal: embeddings must feel like built-in graph statistics (degree, pagerank) and seamlessly compose with queries, transformations, and ML workflows.

All embedding outputs must follow consistent semantics:

(node, layer) -> vector

Vectors must propagate through query results, pipelines, and graph operations.

---

Standardize embedding output schema.

Every embedding operation must produce a structured table with columns:

node
layer
embedding
embedding_dim
method
timestamp

embedding must be a fixed-length vector.

Define canonical dtype:

np.ndarray[float32]

Internally allow:

jax.numpy
torch.tensor

but export must normalize to NumPy.

---

Define a unified embedding result container.

Create:

class EmbeddingResult

Capabilities:

embedding[node]
embedding[(node, layer)]

embedding.vectors
embedding.nodes
embedding.layers
embedding.dimension
embedding.method

Accessors:

embedding.to_numpy()
embedding.to_pandas()
embedding.to_arrow()
embedding.to_parquet()

Similarity operations:

embedding.similarity(a, b)
embedding.distance(a, b)

embedding.most_similar(node, k=10)
embedding.knn(node, k=10)

Batch operations:

embedding.similarity_matrix(nodes)
embedding.subset(nodes)

---

Ensure (node, layer) semantics are consistent.

Define canonical node identity:

(node_id, layer_id)

All embedding models must operate on this tuple.

Provide helper utilities:

flatten_nodes()
expand_layers()
group_by_node()
group_by_layer()

Support both:

node-only embeddings
layer-aware embeddings

If layers are ignored, set:

layer = None

---

Integrate embeddings into the Q DSL.

Add operator:

embed(method, **params)

Example usage:

Q.nodes()
 .embed("node2vec", dimensions=128)
 .order_by("degree")
 .limit(100)
 .execute(net)

Result must include:

embedding

as a column.

Ensure compatibility with:

select
filter
mutate
group_by
order_by
limit

Example:

Q.nodes()
 .embed("node2vec", dim=128)
 .mutate(sim=lambda x: cosine_similarity(x.embedding))

---

Integrate embeddings into graph_ops.

Allow embedding generation in fluent operations:

nodes(net)
    .embed(method="node2vec", dim=64)
    .mutate(cluster=lambda df: kmeans(df.embedding))

Embedding column must appear automatically.

Operations must support:

filter
mutate
aggregate
join

Ensure lazy execution compatibility.

---

Integrate embeddings with pipelines.

Create pipeline step:

NodeEmbedding

File:

py3plex/pipeline/steps/embedding.py

Interface:

NodeEmbedding(
    method="node2vec",
    dimensions=128,
    **kwargs
)

Usage:

Pipeline([
    ("embed", NodeEmbedding(method="node2vec")),
    ("cluster", NodeClustering())
])

Output must pass embedding vectors to subsequent steps.

---

Add embedding caching.

Embeddings must support reproducible caching.

Define hash:

graph_structure_hash
method
parameters
version

Cache key:

embedding_{hash}.parquet

API:

net.embed(..., cache=True)

If cached embedding exists:

load instead of recompute

---

Add embedding provenance tracking.

Each embedding result must store metadata:

embedding.method
embedding.parameters
embedding.graph_hash
embedding.creation_time
embedding.library_version

Provide method:

embedding.info()

Return structured metadata.

---

Implement deterministic replay.

Allow reconstruction:

embedding.reproduce(network)

Use stored parameters.

---

Add vector operations.

Methods:

embedding.normalize()
embedding.reduce(method="pca", dim=2)
embedding.reduce(method="umap", dim=2)
embedding.cluster(method="kmeans", k=10)

Return structured node mappings.

---

Add vector index acceleration.

Build optional ANN index.

Supported backends:

faiss
annoy
hnsw

API:

embedding.build_index(method="hnsw")
embedding.knn(node, k=10)

---

Ensure export compatibility.

Supported export formats:

pandas
arrow
parquet
numpy

Example:

embedding.to_pandas()
embedding.to_arrow()
embedding.save("embeddings.parquet")

Load:

EmbeddingResult.load("embeddings.parquet")

---

Ensure QueryResult compatibility.

Embedding columns must survive:

query execution
pipeline execution
graph transformations
joins
filters

Example:

Q.nodes()
 .embed("node2vec")
 .filter("layer == 'social'")
 .execute(net)

Embedding must persist in result table.

---

Add validation utilities.

Checks:

embedding.dimension_consistency
node_count_match
layer_alignment

Expose:

embedding.validate()

---

Add performance optimizations.

Requirements:

vector storage contiguous
avoid Python loops
use batched operations
support sparse graph traversal

Ensure compatibility with graphs with:

>1M nodes
>10M edges

---

Add test coverage.

Tests:

test_embedding_query_integration
test_embedding_graph_ops
test_embedding_pipeline_step
test_embedding_export_formats
test_embedding_similarity
test_embedding_knn
test_embedding_caching
test_embedding_metadata
test_multiplex_embedding_semantics

---

Add documentation examples.

Examples:

# generate embeddings
emb = net.embed(method="node2vec", dimensions=128)

# similarity search
emb.most_similar(("Alice", "social"))

# clustering
clusters = emb.cluster(method="kmeans", k=10)

# query integration
Q.nodes().embed("node2vec").execute(net)

# pipeline integration
Pipeline([
    ("embed", NodeEmbedding(method="node2vec")),
    ("cluster", NodeClustering())
])

---

Embedding must behave as a core graph primitive fully integrated with the DSL, pipelines, and multilayer semantics of py3plex.

no new md allowed
